Add autoupgrade test for sle15

QingmingSu opened the pull request:
Adopt version switch in autoupgrade and zdup tests by 
moving version switch modules into load_patching_tests and
load_zdup_tests functions to make the version switching
applicable for all kinds of upgrading tests, not only offline
upgrade, but also autoupgrade and zdup.

- Related ticket: https://progress.opensuse.org/issues/30056
- Needles: N/A
- Verification run: 
   * [upgrade] http://openqa-apac1.suse.de/tests/98
   * [autoupgrade] http://openqa-apac1.suse.de/tests/97
   * [zdup] http://openqa-apac1.suse.de/tests/99

rwx788 commented on tests/autoyast/clone.pm:
```diff
@@ -26,8 +26,6 @@ sub run {
     upload_logs '/root/autoinst.xml';
 
-    # original autoyast on kernel cmdline
-    upload_logs '/var/adm/autoinstall/cache/installedSystem.xml';
```
Why do you want to remove this? It's current way of getting profile which was used for installation and bug for SLE15 is already fixed: https://bugzilla.suse.com/show_bug.cgi?id=1075154

QingmingSu replied:
I did get test failure due to the bug, but I thought the installedSystem.xml was renamed to "autoinst.ycp" on sle 15. I should treat it carefully for the potential product issue.
And all logs in /var/adm/ will be uploaded in later test module autoyast/logs.pm, seems it is not
required to upload here.
I will discard my change on autoyast/clone.pm 
Thanks for pointing out the issue.

rwx788 commented on lib/version_utils.pm:
```diff
@@ -119,7 +119,8 @@ sub is_storage_ng {
 
 sub is_sle12_hdd_in_upgrade {
-    return get_var('UPGRADE') && !sle_version_at_least('15', version_variable => 'HDDVERSION');
+    return (get_var('UPGRADE') || get_var('ONLINE_MIGRATION') || get_var('ZDUP') || get_var('AUTOUPGRADE'))
```
Why not to reuse is_upgrade_test method you've introduced here as well?

QingmingSu replied:
The function is_upgrade_test is defined in main_common.pm, I thought it was weird to use main_common.pm function in version_utils.pm 
I will update according to mitiao's suggestion: move is_upgrade_test to version_utils.pm and rename it to is_upgrade, then I can reuse the function here in is_sle12_hdd_in_upgrade

mitiao commented on lib/version_utils.pm:
```diff
@@ -119,7 +119,8 @@ sub is_storage_ng {
 
 sub is_sle12_hdd_in_upgrade {
-    return get_var('UPGRADE') && !sle_version_at_least('15', version_variable => 'HDDVERSION');
+    return (get_var('UPGRADE') || get_var('ONLINE_MIGRATION') || get_var('ZDUP') || get_var('AUTOUPGRADE'))
```
It is duplicate with `is_upgrade_test`

QingmingSu replied:
Hwww, I put duplicated code here because I thought it was weird to "use main_common 'is_upgrade_test';" in vertion_utils.pm

mitiao commented on lib/main_common.pm:
```diff
@@ -884,6 +886,10 @@ sub is_mediacheck {
 
+sub is_upgrade_test {
+    return get_var('UPGRADE') || get_var('ONLINE_MIGRATION') || get_var('ZDUP') || get_var('AUTOUPGRADE');
+}
```
How about move it to `version_utils.pm` and just use `is_upgrade` as function name?

QingmingSu replied:
Yes, sounds good.

QingmingSu replied:
Yes, I will update according to your comment.
Thanks for the good suggestion.

rwx788 commented on data/autoyast_sle15/autoyast_media_up_gnome.xml:
```diff
@@ -0,0 +1,91 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE profile>
+<profile xmlns="http://www.suse.com/1.0/yast2ns" xmlns:config="http://www.suse.com/1.0/configns">
+  <add-on>
```
Please, be aware, that once we upgrade os-autoinst on osd we can use variables expansion, like we do here: https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/4176/files#diff-954e34a9a9a0683ba4ebf7de4129b6b9R7 Hence you may consider switching to registration instead of all packages DVD.

QingmingSu replied:
I think both SCC and all-packages DVD should be tested for autoupgrade.

QingmingSu replied:
BTW, upgrade with scc is not implemented yet at the time. So I only test autoupgrade with all-packages media here.

QingmingSu commented:
Updated:
- move "is_upgrade_test" to version_utils.pm and rename it to is_upgrade, reuse it in "is_sle12_hdd_in_upgrade"
- discard the change on autoyast/clone.pm (it is known bug bsc#1075154 and fixed)

Verification tests are running.

QingmingSu commented:
The change has been verified with new test runs:
   * upgrade: http://openqa-apac1.suse.de/tests/108
   * autoupgrade: http://openqa-apac1.suse.de/tests/109
   But there is a new error reported in autoupgrade (with sle15 daily build 418.3), which is likely a new bug.

@rwx788 Did you notice this failure: http://openqa-apac1.suse.de/tests/109#step/installation/4 ?
I suspect it is a new bug of yast.

okurz commented on data/autoyast_sle15/autoyast_media_up_gnome.xml:
```diff
@@ -0,0 +1,91 @@
+    </errors>
+    <warnings>
+      <show config:type="boolean">true</show>
+      <timeout config:type="integer">10</timeout>
```
In (all?) other profiles @rwx788 switched off the timeout for all kind of messages now so that we explicitly acknowledge all and let no error slip by. Maybe we want the same here?

QingmingSu replied:
Yes, absolutely I want to keep the same change. Updated the xml file to turn off all timeout for all messages.

okurz commented on tests/migration/version_switch_origin_system.pm:
```diff
@@ -18,13 +19,15 @@ use warnings;
+    # Before upgrade or after rollback, switch to original system version
+    # Do NOT use HDDVERSION because it might be changed in another test
+    # module, such as: reboot_and_install.pm
+    my $original_version = get_required_var('ORIGIN_SYSTEM_VERSION');
```
That's a very good choice of variable name, thank you!

QingmingSu commented:
Updated main.pm with following changes to make it possible that not update original system before upgrade:
-        load_patching_tests() if get_var('PATCH');
+        if (get_var('PATCH')) {
+            load_patching_tests();
+        }
+        else {
+            load_boot_tests();
+        }

rwx788 commented:
@QingmingSu I haven't tested upgrade scenarios with autoyast, but it's definitely a bug

QingmingSu commented:
@rwx788 Then I will report a bug against the failure.

QingmingSu commented:
Verified again on latest build 419.1, offline upgrade passed as expected (http://openqa-apac1.suse.de/tests/118), but autoupgrade still failed with (bsc#1075744) (http://openqa-apac1.suse.de/tests/119)

DimStar77 commented:
Breaks all tw zdup tests, e.g. https://openqa.opensuse.org/tests/583028#step/version_switch_upgrade_target/1

QingmingSu commented:
Oops, I will fix that issue for opensuse ASAP. @DimStar77

QingmingSu commented:
PR#4212 has been submitted to fix the issue for opensuse. Sorry for inconvenience. @DimStar77
